files/acl.py

```diff
@@ -151,69 +145,87 @@ def split_entry(entry):
     else:
         t = None
 
-    p = normalize_permissions(p)
+    return [d, t, e, p]
 
-    return [d,t,e,p]
 
-def get_acls(module,path,follow):
+def build_entry(etype, entity, permissions=None):
+    '''Builds and returns an entry string. Does not include the permissions bit if they are not provided.'''
+    return etype + ':' + entity + (':' + permissions if permissions else '')
 
-    cmd = [ module.get_bin_path('getfacl', True) ]
-    if not follow:
-        cmd.append('-h')
-    # prevents absolute path warnings and removes headers
-    cmd.append('--omit-header')
-    cmd.append('--absolute-names')
-    cmd.append(path)
 
-    return _run_acl(module,cmd)
+def build_command(module, mode, path, follow, default, recursive, entry=''):
+    '''Builds and returns agetfacl/setfacl command.'''
+    if mode == 'set':
+        cmd = [module.get_bin_path('setfacl', True)]
+        cmd.append('-m "%s"' % entry)
+    elif mode == 'rm':
+        cmd = [module.get_bin_path('setfacl', True)]
+        cmd.append('-x "%s"' % entry)
+    else:  # mode == 'get'
+        cmd = [module.get_bin_path('getfacl', True)]
+        cmd.append('--omit-header')
+        cmd.append('--absolute-names')
 
-def set_acl(module,path,entry,follow,default):
+    if recursive:
+        cmd.append('--recursive')
 
-    cmd = [ module.get_bin_path('setfacl', True) ]
     if not follow:
         cmd.append('-h')
+
     if default:
-        cmd.append('-d')
-    cmd.append('-m "%s"' % entry)
-    cmd.append(path)
+        if(mode == 'rm'):
+            cmd.append('-k')
+        else:  # mode == 'set' or mode == 'get'
+            cmd.append('-d')
 
-    return _run_acl(module,cmd)
+    cmd.append(path)
+    return cmd
 
-def rm_acl(module,path,entry,follow,default):
 
-    cmd = [ module.get_bin_path('setfacl', True) ]
-    if not follow:
-        cmd.append('-h')
-    if default:
-        cmd.append('-k')
-    entry = entry[0:entry.rfind(':')]
-    cmd.append('-x "%s"' % entry)
-    cmd.append(path)
+def acl_changed(module, cmd):
+    '''Returns true if the provided command affects the existing ACLs, false otherwise.'''
+    cmd = cmd[:]  # lists are mutables so cmd would be overriden without this
+    cmd.insert(1, '--test')
+    lines = run_acl(module, cmd)
+    return not all(line.endswith('*,*') for line in lines)
```

Comment:
Hi @puiterwijk, as I was building my defense for this syntax, I noticed that `all()` was Python 2.5+ only, so I replaced it with an ugly `for`-loop in f8027a7. Hope this looks better to you.

About the original syntax though, I do believe (and I checked with a very experienced Python dev colleague of mine) that this was the easiest way to put/read this.
In fact, this makes much more sense if you read it out loud, as this translates in English to:

```
if not all lines end with '*,*' then acls have changed, otherwise acls haven't changed
```

On top of that, I was using a generator so it can cleanly exit prematurely, for efficiency. Not exactly sure what difference that makes here, but assuming one is running the `--recursive` option on a folder of hundred thousands sub-folders and files, on old hardware (after all, we do support Python 2.4 :-) ), it's always nice to terminate early if we can. My `for`-loop keeps this behavior.